spanconfig: implement the ProtectedTSReader interface on the KVSubscriber

arulajmani opened the pull request:
This patch makes the `KVSubscriber` implement the `ProtectedTSReader`
interface. This allows the GC queue to consult protected timestamp
information reconciled by sql tenants when making GC decisions.

We also initialize the `spanconfigptsreader.adapter` struct with a
KVSubscriber now. The adapter was conceived with the intention to
encapsulate PTS information from both the old and new subsystem for
release 22.1 behind a single source for the GC queue to consume.
Threading in the KVSubscriber actually achieves this. We return the
minimum freshness of the cache and KVSubscriber to the GC queue from
the adapter.

Release justification: bug fixes and low risk updates to new
functionality

Release note: None

arulajmani committed:
spanconfig: make KVSubscriber implement the ProtectedTSReader interface

This patch makes the `KVSubscriber` implement the `ProtectedTSReader`
interface. This allows the GC queue to consult protected timestamp
information reconciled by sql tenants when making GC decisions.

We also initialize the `spanconfigptsreader.adapter` struct with a
KVSubscriber now. The adapter was conceived with the intention to
encapsulate PTS information from both the old and new subsystem for
release 22.1 behind a single source for the GC queue to consult.
Threading in the KVSubscriber actually achieves this. We return the
minimum freshness of the cache and KVSubscriber to the GC queue as
the "combined" freshnsess of the adapter.

Release justification: bug fixes and low risk updates to new
functionality

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/77338)
<!-- Reviewable:end -->

arulajmani commented:
@irfansharif made the changes you mentioned on Slack, namely, removed `Start` from the `KVSubscriber` interface and moved `ForEachOverlappingSpanConfig` from the interface (`spanconfig.Store`) to the concrete impl. PTAL.

irfansharif commented on pkg/kv/kvserver/replica_protected_timestamp_test.go:
```diff
@@ -145,7 +150,9 @@ func TestCheckProtectedTimestampsForGC(t *testing.T) {
-				canGC, _, gcTimestamp, oldThreshold, newThreshold := r.checkProtectedTimestampsForGC(ctx, makeTTLDuration(10))
+				canGC, _, gcTimestamp, oldThreshold, newThreshold, err := r.checkProtectedTimestampsForGC(ctx, makeTTLDuration(10))
+				require.NoError(t, err)
+				require.True(t, canGC)
```
Duplicate line?

irfansharif reviewed (approved):
LGTM modulo [e,e) showing up in the test grammar to represent a point span. It’s an invalid span if e is both inclusive and exclusive. In our test I think we’ll get the same coverage just testing for the single alphabet span [e,f). If you really want to test keys (not necessary IMO, KV doesn’t ask for configs for keys, does it?) it’s worth adding something in the testing grammar to represent it. Either [e,nil) or some syntax for point specifically. I prefer [e,f) also because at various places we assert on spans having non-nil end keys.

arulajmani reviewed (commented):
TFTR! I removed the point tests considering KV doesn't ask for them and given your comment. The CI failure seems to be a flake, going to press merge. 

bors r+

> Dismissed @irfansharif from a discussion.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/77338#-:-MxLdHRR0ReQYlpK60vS:b-f3hovk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, and @nvanbenschoten)

___
*[pkg/kv/kvserver/replica_protected_timestamp_test.go, line 155 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/77338#-MxGSoQ7zZiSVurEnUTA-r2-155:-MxLdGZNCUqasVMcn3JT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d37617f6cb12418b06a592e8d4801e2e09e0d652/pkg/kv/kvserver/replica_protected_timestamp_test.go#L155)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Duplicate line?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->
